proposals/0031-hlsl-vector-matrix-operations.md

```diff
@@ -210,10 +205,516 @@ void __builtin_VectorAccumulate(vector<TY, NUM> InputVector,
+  DATA_TYPE_UINT16 = 3,           // ComponentType::U16
+  DATA_TYPE_SINT32 = 4,           // ComponentType::I32
+  DATA_TYPE_UINT32 = 5,           // ComponentType::U32
+  DATA_TYPE_FLOAT16 = 8,          // ComponentType::F16
+  DATA_TYPE_FLOAT32 = 9,          // ComponentType::F32
+  DATA_TYPE_SINT8_T4_PACKED = 17, // ComponentType::PackedS8x32
+  DATA_TYPE_UINT8_T4_PACKED = 18, // ComponentType::PackedU8x32
+  DATA_TYPE_UINT8 = 19,           // ComponentType::U8
+  DATA_TYPE_SINT8 = 20,           // ComponentType::I8
+  DATA_TYPE_FLOAT8_E4M3 = 21,     // ComponentType::F8_E4M3
+                                  // (1 sign, 4 exp, 3 mantissa bits)
+  DATA_TYPE_FLOAT8_E5M2 = 22,     // ComponentType::F8_E5M2
+                                  // (1 sign, 5 exp, 2 mantissa bits)
+};
+
+} // namespace linalg
+} // namespace dx
+```
+
+See the Type Interpretations section of [Proposal 0029] for more information.
+
+### enum MatrixLayout
+
+The `dx::linalg::MatrixLayout` enum defines the different possible layouts of a
+matrix in memory.
+
+```c++
+namespace dx {
+namespace linalg {
+
+enum MatrixLayout {
+  MATRIX_LAYOUT_ROW_MAJOR = 0,
+  MATRIX_LAYOUT_COLUMN_MAJOR = 1,
+  MATRIX_LAYOUT_MUL_OPTIMAL = 2,
+  MATRIX_LAYOUT_OUTER_PRODUCT_OPTIMAL = 3
+};
+
+} // namespace linalg
+} // namespace dx
+```
+
+See the Matrix Layouts section of [Proposal 0029] for more information.
+
+### struct MatrixRef / RWMatrixRef
+
+`dx::linalg::MatrixRef` and `dx::linalg::RWMatrixRef` specify a reference to a
+matrix in memory, including information about the dimensions, layout and
+numerical format of the matrix. Some of this information must be known at
+compile time - these are captured in template parameters - while others can be
+determined at runtime, and are therefore members of the struct.
+
+Since the `ByteAddressBuffer` and `RWByteAddressBuffer` are not convertable to
+each other, it is neccesary to use different named types for the matrix
+references. A base class, `dx::linalg::MatrixRefImpl` is used for the common
+code.
+
+Other functions in the API are implemented in terms of
+`dx::linalg::MatrixRefImpl`. An option considered was to use a template
+parameter for the entire matrix type. However, this approach results in hard to
+understand error messages if a non-MatrixRef type is passed for that parameter.
+
+Example usage:
+
+```c++
+ByteAddressBuffer ROBuffer;
+RWByteAddressBuffer RWBuffer;
+
+void Example() {
+  using namespace dx::linalg;
+
+  MatrixRef<DATA_TYPE_FLOAT16, 4, 4, MATRIX_LAYOUT_MUL_OPTIMAL, true> MatrixA =
+      {ROBuffer, /*offset=*/128, /*stride=*/0};
+
+  MatrixRef<DATA_TYPE_FLOAT16, 4, 4, MATRIX_LAYOUT_ROW_MAJOR, true>
+      MatrixB = {ROBuffer, /*offset=*/128, /*stride=*/16};
+
+  RWMatrixRef<DATA_TYPE_FLOAT16, 128, 256, MATRIX_LAYOUT_OUTER_PRODUCT_OPTIMAL>
+      MatrixC = {RWBuffer, /*offset=*/64, /*stride=*/0};
+}
+```
+
+Template parameters:
+
+- `DT` - the type used to store elements of the matrix in memory
```

Comment:
minor nit that I think it would be helpful to annotate the parameters/members with their data types. I know you can look below at the implementation to see but I think it would improve readability to duplicate here. Feel free to tell me you disagree though.... 😄